lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -797,3 +971,110 @@ String _maskFilterToCss(ui.MaskFilter maskFilter) {
+      '<feFlood flood-color="${colorToCssString(color)}" flood-opacity="1" result="flood">'
+      '</feFlood>'
+      '<feComposite in="flood" in2="SourceGraphic" operator="out" result="comp">'
+      '</feComposite>'
+      '</filter></svg>';
+}
+
+String _xorColorFilterToSvg(ui.Color color) {
+  _filterIdCounter += 1;
+  return '<svg width="0" height="0">'
+      '<filter id="_fcf$_filterIdCounter" '
+      'filterUnits="objectBoundingBox" x="0%" y="0%" width="100%" height="100%">'
+      '<feFlood flood-color="${colorToCssString(color)}" flood-opacity="1" result="flood">'
+      '</feFlood>'
+      '<feComposite in="flood" in2="SourceGraphic" operator="xor" result="comp">'
+      '</feComposite>'
+      '</filter></svg>';
+}
+
+// The source image and color are composited using :
+// result = k1 *in*in2 + k2*in + k3*in2 + k4.
+String _compositeColorFilterToSvg(ui.Color color, double k1, double k2, double k3 , double k4) {
+  _filterIdCounter += 1;
+  return '<svg width="0" height="0">'
+      '<filter id="_fcf$_filterIdCounter" '
+      'filterUnits="objectBoundingBox" x="0%" y="0%" width="100%" height="100%">'
+      '<feFlood flood-color="${colorToCssString(color)}" flood-opacity="1" result="flood">'
+      '</feFlood>'
+      '<feComposite in="flood" in2="SourceGraphic" '
+      'operator="arithmetic" k1="$k1" k2="$k2" k3="$k3" k4="$k4" result="comp">'
+      '</feComposite>'
+      '</filter></svg>';
+}
+
+// Porter duff source * destination , keep source alpha.
+// First apply color filter to source to change it to [color], then
+// composite using multiplication.
+String _modulateColorFilterToSvg(ui.Color color) {
+  _filterIdCounter += 1;
+  final double r = color.red / 255.0;
+  final double b = color.blue / 255.0;
+  final double g = color.green / 255.0;
+  return '<svg width="0" height="0">'
+      '<filter id="_fcf$_filterIdCounter" '
+      'filterUnits="objectBoundingBox" x="0%" y="0%" width="100%" height="100%">'
+      '<feColorMatrix values="0 0 0 0 $r ' // Ignore input, set it to absolute.
+      '0 0 0 0 $g '
+      '0 0 0 0 $b '
+      '0 0 0 1 0" result="recolor"/>'
+      '<feComposite in="recolor" in2="SourceGraphic" '
+      'operator="arithmetic" k1="1" k2="0" k3="0" k4="0" result="comp">'
+      '</feComposite>'
+      '</filter></svg>';
+}
+
+// Uses feBlend element to blend source image with a color.
+String _blendColorFilterToSvg(ui.Color color, String feBlend,
+    {bool swapLayers = false}) {
+  _filterIdCounter += 1;
+  return '<svg width="0" height="0">'
+      '<filter id="_fcf$_filterIdCounter" filterUnits="objectBoundingBox" '
+      'x="0%" y="0%" width="100%" height="100%">'
+      '<feFlood flood-color="${colorToCssString(color)}" flood-opacity="1" result="flood">'
+      '</feFlood>' +
+      (swapLayers
+      ? '<feBlend in="SourceGraphic" in2="flood" mode="$feBlend"/>'
+      : '<feBlend in="flood" in2="SourceGraphic" mode="$feBlend"/>') +
+      '</filter></svg>';
```

Comment:
These SVG snippets contain a lot of static content that need to be parsed from HTML. Wouldn't using the SVG API be more efficient?

Reply:
Didn't want to pull in svg library, not as big as dart:html but not small.

Reply:
Something tells me we already have it all included :)

It's OK. We can profile it later and see what the size/speed trade-off is like.